### PR TITLE
Provide bt result

### PR DESCRIPTION
**Pull request by mradams-bastian**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | [2991](https://github.com/ros-planning/navigation2/issues/2991) |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | Gazebo |

---

## Description of contribution in a few bullet points

Added input parameter 'final_bt_status' of type nav2_behavior_tree::BtStatus to the goalCompleted() function of navigator.hpp that makes the final status of the behavior tree run available to the navigator. This could provide the navigator with important context while populating the result and doing anything else it might do during that function.

None of the navigators that are included in nav2 currently do anything with this goalCompleted() function, but this could be a nice feature for developers who are creating their own custom navigators, and the current nav2 navigators may also benefit from this feature in the future. For example, a parameter in the result message could provide some context to the client as to whether an aborted result corresponds to a preemption or a behavior tree failure.

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

**Comment by padhupradheep** [comment]
To give a heads up, there are some linter issues, that you might want to fix.. see the details in the release test. You can also get those results by simply running `ament_uncrustify` and `ament_cpplint`

File: nav2_bt_navigator/src/navigators/navigate_to_pose.cpp [hunk 1]
```diff
@@ -101,7 +101,8 @@ NavigateToPoseNavigator::goalReceived(ActionT::Goal::ConstSharedPtr goal)
 }
 
 void
-NavigateToPoseNavigator::goalCompleted(typename ActionT::Result::SharedPtr /*result*/)
+NavigateToPoseNavigator::goalCompleted(typename ActionT::Result::SharedPtr /*result*/,
+                                       nav2_behavior_tree::BtStatus /*final_bt_status*/)
```
Comment (SteveMacenski) [inline nav2_bt_navigator/src/navigators/navigate_to_pose.cpp:105]: This is not linted properly, there are no line alignments in C++ (that's python), this should just be a new line with 2 space indent

File: nav2_bt_navigator/src/navigators/navigate_through_poses.cpp [hunk 2]
```diff
@@ -87,7 +87,8 @@ NavigateThroughPosesNavigator::goalReceived(ActionT::Goal::ConstSharedPtr goal)
 }
 
 void
-NavigateThroughPosesNavigator::goalCompleted(typename ActionT::Result::SharedPtr /*result*/)
+NavigateThroughPosesNavigator::goalCompleted(typename ActionT::Result::SharedPtr /*result*/,
+                                             nav2_behavior_tree::BtStatus /*final_bt_status*/)
```
Comment (SteveMacenski) [inline nav2_bt_navigator/src/navigators/navigate_through_poses.cpp:91]: Ditto

File: nav2_bt_navigator/src/navigators/navigate_to_pose.cpp (hunk 1 above)
Comment (SteveMacenski) [reply nav2_bt_navigator/src/navigators/navigate_to_pose.cpp:105]: Also, make BTStatus `const`

File: nav2_bt_navigator/src/navigators/navigate_through_poses.cpp (hunk 2 above)
Comment (SteveMacenski) [reply nav2_bt_navigator/src/navigators/navigate_through_poses.cpp:91]: Also, make BTStatus const

File: nav2_bt_navigator/include/nav2_bt_navigator/navigators/navigate_to_pose.hpp [hunk 3]
```diff
@@ -106,8 +106,11 @@ class NavigateToPoseNavigator
    * @brief A callback that is called when a the action is completed, can fill in
    * action result message or indicate that this action is done.
    * @param result Action template result message to populate
+   * @param final_bt_status Resulting status of the behavior tree execution that may be
+   * referenced while populating the result.
    */
-  void goalCompleted(typename ActionT::Result::SharedPtr result) override;
+  void goalCompleted(typename ActionT::Result::SharedPtr result,
+                     nav2_behavior_tree::BtStatus final_bt_status) override;
```
Comment (SteveMacenski) [inline nav2_bt_navigator/include/nav2_bt_navigator/navigators/navigate_to_pose.hpp:113]: Indent + const

File: nav2_bt_navigator/include/nav2_bt_navigator/navigators/navigate_through_poses.hpp [hunk 4]
```diff
@@ -95,8 +95,11 @@ class NavigateThroughPosesNavigator
    * @brief A callback that is called when a the action is completed, can fill in
    * action result message or indicate that this action is done.
    * @param result Action template result message to populate
+   * @param final_bt_status Resulting status of the behavior tree execution that may be
+   * referenced while populating the result.
    */
-  void goalCompleted(typename ActionT::Result::SharedPtr result) override;
+  void goalCompleted(typename ActionT::Result::SharedPtr result,
+                     nav2_behavior_tree::BtStatus final_bt_status) override;
```
Comment (SteveMacenski) [inline nav2_bt_navigator/include/nav2_bt_navigator/navigators/navigate_through_poses.hpp:102]: Indent + const

File: nav2_bt_navigator/include/nav2_bt_navigator/navigator.hpp [hunk 5]
```diff
@@ -292,10 +293,11 @@ class Navigator
   virtual void onPreempt(typename ActionT::Goal::ConstSharedPtr goal) = 0;
 
   /**
-   * @brief A callback that is called when a the action is completed, can fill in
+   * @brief A callback that is called when a the action is completed; Can fill in
    * action result message or indicate that this action is done.
    */
-  virtual void goalCompleted(typename ActionT::Result::SharedPtr result) = 0;
+  virtual void goalCompleted(typename ActionT::Result::SharedPtr result,
+                             nav2_behavior_tree::BtStatus final_bt_status) = 0;
```
Comment (SteveMacenski) [inline nav2_bt_navigator/include/nav2_bt_navigator/navigator.hpp:300]: Also, make BTStatus const

File: nav2_bt_navigator/include/nav2_bt_navigator/navigator.hpp [hunk 6]
```diff
@@ -265,12 +265,13 @@ class Navigator
   }
 
   /**
-   * @brief An intermediate compution function to mux navigators
+   * @brief An intermediate completion function to mux navigators
    */
-  void onCompletion(typename ActionT::Result::SharedPtr result)
+  void onCompletion(typename ActionT::Result::SharedPtr result,
+                    nav2_behavior_tree::BtStatus final_bt_status)
```
Comment (SteveMacenski) [inline nav2_bt_navigator/include/nav2_bt_navigator/navigator.hpp:271]: Also, make BTStatus const

File: nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server.hpp [hunk 7]
```diff
@@ -40,7 +40,8 @@ class BtActionServer
   typedef std::function<bool (typename ActionT::Goal::ConstSharedPtr)> OnGoalReceivedCallback;
   typedef std::function<void ()> OnLoopCallback;
   typedef std::function<void (typename ActionT::Goal::ConstSharedPtr)> OnPreemptCallback;
-  typedef std::function<void (typename ActionT::Result::SharedPtr)> OnCompletionCallback;
+  typedef std::function<void (typename ActionT::Result::SharedPtr,
+      nav2_behavior_tree::BtStatus)> OnCompletionCallback;
```
Comment (SteveMacenski) [inline nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server.hpp:44]: Just 2 spaces

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
LGTM, I don't usually mention by hand the linting errors, but I want to get this done ASAP so I can release this afternoon

**Comment by mradams-bastian** [comment]
K, I'll get those fixed soon

**Comment by SteveMacenski** [comment]
Toggling for CI, it didn't trigger for some reason

**Review by SteveMacenski — APPROVED** [review]
Just waiting on CI

**Comment by SteveMacenski** [comment]
Thanks @mradams-bastian

**Pull request merged**